### PR TITLE
Hooks use_state, use_effect, use_async

### LINKED ISSUE #37: Hooks
We want Hooks.

The pyedifice API is based on the old legacy React Components. I can see why React Hooks was adopted — the legacy Components API is really bad.

We especially want  an async [`useEffect`](https://legacy.reactjs.org/docs/hooks-effect.html).

The async `useEffect` Hook would have four important features which are bothersome to implement manually over and over.

1. Runs effect only when props change.
2. Cancels the effect if the Component is unmounted before the effect runs.
3. Cancels the effect if another effect starts before this one finishes.
4. Runs an additional cleanup effect.

Idea: implement Hooks as a [`with` statement context manager](https://docs.python.org/3/reference/datamodel.html#context-managers)

See also [PureScript `useAff`](https://pursuit.purescript.org/packages/purescript-react-basic-hooks/8.2.0/docs/React.Basic.Hooks.Aff#v:useAff)

### PR DESCRIPTION
**Pull request by jamesdbrock**
image

image

image

Resolves #37

### CODE REVIEW

**Comment by jamesdbrock** [comment]
What about unwind changes when exception is thrown?

File: edifice/_component.py [hunk 1]
```diff
@@ -443,6 +493,200 @@ def set_state(self, **kwargs):
...
+        cleans itself up after exceptions.
+        Make sure that the :code:`fn_coroutine` function
+        does not try to do anything with this Component after an
+        :code:`asyncio.CancelledError`
+        is raised, because this Component may at that time
+        already have been unmounted.
+
+        Example::
+
+            def render():
```
Comment (considerate) [inline edifice/_component.py:633]:
```suggestion
            def render(self):
```

File: edifice/_component.py [hunk 2]
```diff
@@ -443,6 +493,200 @@ def set_state(self, **kwargs):
...
+        dependencies: tp.Any,
+    ) -> None:
+        """
+        Effect Hook for use inside the :func:`render` function.
+
+        Behaves like `React useEffect <https://react.dev/reference/react/useEffect>`_.
+
+        Example::
+
+            def render():
```
Comment (considerate) [inline edifice/_component.py:552]:
```suggestion
            def render(self):
```

File: edifice/engine.py [hunk 3]
```diff
@@ -395,6 +410,10 @@ def _render(self, component: Component, render_context: _RenderContext):
             render_context.component_parent = component._edifice_internal_parent
             return ret
 
+		# Before the render, set the hooks indices of the Component to 0.
```
Comment (considerate) [inline edifice/engine.py:413]:
```suggestion
        # Before the render, set the hooks indices of the Component to 0.
```

**Comment by jamesdbrock** [comment]
Because these primitive Hooks are Component instance methods and have to be called on `self`, I think it's impossible for the user to create new custom Hooks derived from these primitive Hooks. So maybe we need a new approach entirely.

**Comment by considerate** [comment]
@jamesdbrock That's a really good catch. I think they might have to go on the render context after all.

I'd be in favor of merging this first and doing the move to the render context in a different PR. What do you say?

**Comment by jamesdbrock** [comment]
> I think they might have to go on the render context after all.

If they are on the `_RenderContext` then how will Hook state be transmitted across renders?

https://github.com/pyedifice/pyedifice/blob/b5438930b8cea00643f50ed9d4eacb5bde42fad0/edifice/engine.py#L59

> I'd be in favor of merging this first and doing the move to the render context in a different PR. What do you say?

I think we should concentrate on #53 and think about how Hooks can be built on that PR.

**Comment by considerate** [comment]
> If they are on the _RenderContext then how will Hook state be transmitted across renders?

Hmm, maybe they need to go on the engine object. The _RenderContext gets created anew every frame so putting them there doesn't look like the state could be persisted between frames.

**Pull request closed without merging**